stripe/src/main/java/com/stripe/android/net/PollingNetworkHandler.java

```diff
@@ -0,0 +1,188 @@
+package com.stripe.android.net;
+
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.Looper;
+import android.os.Message;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.stripe.android.exception.PollingFailedException;
+import com.stripe.android.exception.StripeException;
+import com.stripe.android.model.Source;
+import com.stripe.android.model.SourceRedirect;
+
+/**
+ * Class to handle polling on a background thread.
+ */
+class PollingNetworkHandler {
+
+    private static final String AUTHORIZATION_FAILED = "The redirect authorization request failed.";
+    private static final long DEFAULT_TIMEOUT_MS = 10000L;
+    private static final int INITIAL_DELAY_MS = 100;
+    private static final String POLLING_EXPIRED = "The polling request has expired.";
+    private static final int POLLING_MULTIPLIER = 2;
+
+    private static final int SUCCESS = 1;
+    private static final int PENDING = 2;
+    private static final int FAILURE = 3;
+    private static final int ERROR = -1;
+    private static final int EXPIRED = -2;
+
+    private final String mClientSecret;
+    private final String mPublishableKey;
+    private final String mSourceId;
+
+    private final long mTimeoutMs;
+    private final boolean mIsInSingleThreadMode;
+    @NonNull private final SourceRetriever mSourceRetriever;
+
+    private Handler mNetworkHandler;
+    private Handler mUiHandler;
+
+    private final Runnable pollRunnable = new Runnable() {
+        @Override
+        public void run() {
+            Message message = null;
+            try {
+                Source source = mSourceRetriever.retrieveSource(
+                        mSourceId,
+                        mClientSecret,
+                        mPublishableKey);
+                if (source.getRedirect() != null) {
+                    switch (source.getRedirect().getStatus()) {
+                        case SourceRedirect.SUCCEEDED:
+                            message = mUiHandler.obtainMessage(SUCCESS);
+                            break;
+                        case SourceRedirect.PENDING:
+                            message = mUiHandler.obtainMessage(PENDING);
+                            break;
+                        case SourceRedirect.FAILED:
+                            message = mUiHandler.obtainMessage(FAILURE);
+                            break;
+                    }
+                }
+            } catch (StripeException stripeEx) {
+                message = mUiHandler.obtainMessage(ERROR, stripeEx);
+            } finally {
+                if (message != null) {
+                    mUiHandler.sendMessage(message);
+                }
+            }
+        }
+    };
+
+    PollingNetworkHandler(@NonNull final String sourceId,
```

Comment:
You keep polling at the same interval if nothing is wrong, but the status is still pending? Okay -- I thought that was where the exponential backoff came in.

So if it fails for a reason other than expiration or timeout, *then* we start backing off exponentially, but otherwise we'll just do a linear polling time?

That's why my initial polling was 100 Millis (actually, 200, because it gets doubled every time).